### PR TITLE
Set ready status for span after inserting segments

### LINKED ISSUE #644: Missing segments in the newly detected span
hastic-app got empty list of detected segments for newly detected span because of `READY` status writes to data base before inserting new segments.

### PR DESCRIPTION
**Pull request by amper43**
fixes #644 

## Problem
Race condition when segments list is large: panel waits `READY` status and queries segments after that but database didn't save them yet.

## Changes

- set `READY` status for `DetectionSpan` **after** inserting segments.

### CODE REVIEW

File: server/src/controllers/analytics_controller.ts [hunk 1]
```diff
@@ -277,14 +277,20 @@ export async function runLearning(id: AnalyticUnit.AnalyticUnitId, from?: number
 
 export async function runDetect(id: AnalyticUnit.AnalyticUnitId, from?: number, to?: number) {
   let previousLastDetectionTime: number = undefined;
+  let range: TimeRange;
+  let intersection = 0;
+
+  const old_cache = await AnalyticUnitCache.findById(id);
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
  const oldCache = await AnalyticUnitCache.findById(id);
```

File: server/src/controllers/analytics_controller.ts [hunk 2]
```diff
@@ -277,14 +277,20 @@ export async function runLearning(id: AnalyticUnit.AnalyticUnitId, from?: number
...
+  if(old_cache !== null) {
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
  if(oldCache !== null) {
```

File: server/src/controllers/analytics_controller.ts [hunk 3]
```diff
@@ -277,14 +277,20 @@ export async function runLearning(id: AnalyticUnit.AnalyticUnitId, from?: number
...
+    intersection = old_cache.getIntersection();
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]:
```suggestion
    intersection = oldCache.getIntersection();
```

File: server/src/controllers/analytics_controller.ts [hunk 4]
```diff
@@ -335,10 +333,26 @@ export async function runDetect(id: AnalyticUnit.AnalyticUnitId, from?: number,
...
+        range.from + intersection,
+        range.to - intersection,
+        Detection.DetectionStatus.READY
+      )
+    );
     await AnalyticUnit.setStatus(id, AnalyticUnit.AnalyticUnitStatus.READY);
   } catch(err) {
     let message = err.message || JSON.stringify(err);
     await AnalyticUnit.setStatus(id, AnalyticUnit.AnalyticUnitStatus.FAILED, message);
+    await Detection.insertSpan(
```
Comment (rozetko) [inline server/src/controllers/analytics_controller.ts]: span should be inserted before status

File: server/src/controllers/analytics_controller.ts [hunk 5]
```diff
@@ -309,23 +312,13 @@ export async function runDetect(id: AnalyticUnit.AnalyticUnitId, from?: number,
...
-    await Detection.insertSpan(
-      new Detection.DetectionSpan(
-        id,
-        range.from + intersection,
-        range.to - intersection,
-        Detection.DetectionStatus.READY
-      )
-    );
+    // TODO: should we use intersection from new or old cache?
+    intersection = cache.getIntersection();
```
Comment (amper43) [inline server/src/controllers/analytics_controller.ts]: let's delete it

**Pull request merged**